zkevm-circuits/src/evm_circuit/op_execution.rs

```diff
@@ -558,6 +608,103 @@ impl<F: FieldExt> OpExecutionGadget<F> {
                     core_state,
                     execution_step,
                 )?,
+                (_, OpcodeId::DUP1) => self.dup1_gadget.assign(
```

Comment:
I think you could have here only one dup gadget the way we have it for push. You would need to change the match statement a bit to handle dup and push cases. But then a problem appears in dup.rs:

vec![Lookup::BusMappingLookup(BusMappingLookup::Stack {
                    index_offset: (DUPX - usize::from(OpcodeId::DUP1.as_u8()))
                        as i32,
                    value: word.expr(),
                    is_write: false,
                })],

We don't have DUPX here anymore, but talked with @han0110 a bit about this and he suggested you could change index_offset to be an expression and then use opcode.expr() instead of DUPX. Would be that ok for you?